pandas/core/frame.py

```diff
@@ -5134,22 +5132,60 @@ def reset_index(  # type: ignore[misc]
     @overload
     def reset_index(
         self,
-        level: Optional[Union[Hashable, Sequence[Hashable]]] = ...,
-        drop: bool = ...,
-        inplace: Literal[True] = ...,
+        level: Optional[Union[Hashable, Sequence[Hashable]]],
+        drop: bool,
+        inplace: Literal[True],
         col_level: Hashable = ...,
         col_fill: Hashable = ...,
     ) -> None:
         ...
 
+    @overload
     def reset_index(
         self,
-        level: Optional[Union[Hashable, Sequence[Hashable]]] = None,
-        drop: bool = False,
-        inplace: bool = False,
-        col_level: Hashable = 0,
-        col_fill: Hashable = "",
+        *,
+        drop: bool,
+        inplace: Literal[True],
+        col_level: Hashable = ...,
+        col_fill: Hashable = ...,
+    ) -> None:
+        ...
+
+    @overload
+    def reset_index(
+        self,
+        *,
+        level: Optional[Union[Hashable, Sequence[Hashable]]],
+        inplace: Literal[True],
+        col_level: Hashable = ...,
+        col_fill: Hashable = ...,
+    ) -> None:
+        ...
+
+    @overload
+    def reset_index(
+        self,
+        *,
+        inplace: Literal[True],
+        col_level: Hashable = ...,
+        col_fill: Hashable = ...,
+    ) -> None:
+        ...
+
+    @overload
+    def reset_index(
+        self,
+        level: Optional[Union[Hashable, Sequence[Hashable]]] = ...,
+        drop: bool = ...,
+        inplace: bool = ...,
+        col_level: Hashable = ...,
+        col_fill: Hashable = ...,
     ) -> Optional[DataFrame]:
+        ...
+
+    def reset_index(
+        self, level=None, drop=False, inplace=False, col_level=0, col_fill=""
```

Comment:
removing the type annotations from the function means that we lose the type checking for the body of the function.

so if you reveal_type on say `level` you will get `Revealed type is 'Any'`

IIUC correctly, the type annotations on the function are used to check the function itself, while the overloads are used when checking other functions that call this one.

so both are needed.